### PR TITLE
Observation analyzer

### PR DESCRIPTION
**Pull request by xjules**
This adds a first suggestion of how the misfits, ie. `(obs - response)` discrepancy could be visualized. It gets its own view, call Observation Analyzer, which could be later reused to apply user-defined transformations on top of observations.
Resolves equinor/ert#766

### CODE REVIEW

**Comment by oysteoh** [comment]
i would encourage you to do a rebase on master since the `data_url's` now are gone and replaced with the actual `data`

**Comment by xjules** [comment]
Current version of misfits, comments / suggestions are welcomed:

image

**Comment by ManInFez** [comment]
The have asked for a sorted histogram showing the summarized misfit for each realization. Is that possible in your solution?

File: ertviz/assets/ert-style.json [hunk 1]
```diff
@@ -29,33 +29,65 @@
...
-                "dash" : "dash"
+                "color": "rgba(56,108,176,0.8)",
+                "dash": "dash"
             },
-            "marker" : null
+            "marker": null
         }
     }
-}
+}
```
Comment (ManInFez) [inline ertviz/assets/ert-style.json:93]: no new line

**Comment by xjules** [comment]
> The have asked for a sorted histogram showing the summarized misfit for each realization. Is that possible in your solution?

No, it's not. Do you have any details. With @hnformentin we also discussed boxplots, which doesn't look that bad:
image

**Comment by hnformentin** [comment]
> > The have asked for a sorted histogram showing the summarized misfit for each realization. Is that possible in your solution?
> 
> No, it's not. Do you have any details? With @hnformentin we also discussed the boxplot, which doesn't look that bad:
> image

I think we talked also to possibly have this plot with the misfit as real value in x-axis (instead of absolute value). Maybe this is easier to interpret. 

I think the summarised misfit would be possible to plot in a histogram for some time series (e.g. of bottom hole pressure, water and oil production rate), where each of these time series can be called data types.
 
The summarized misfit can be the a multivariate misfit measure such as sum of [(f(x_i_t)-z_i_t)/sd_i_t]² for t in the range of the time steps that the data type is measured; where f(x_i_t) is the simulation result, z_i_t the observed data, and sd_i_t the standard deviation for the data type i and time step t. 
(there is probably a better way to format equations here :sweat_smile: )

What do you think about it?

**Comment by xjules** [comment]
> > The have asked for a sorted histogram showing the summarized misfit for each realization. Is that possible in your solution?
> 
> No, it's not. Do you have any details? With @hnformentin we also discussed the boxplot, which doesn't look that bad:
> image

Here comes a new version:
image

**Comment by xjules** [comment]
I've made the misfits squared root and signed:
image

The summary misfits show as Histogram (as suggested):
image

**Comment by oysteoh** [comment]
when choosing a response without observations - and then clicking summary you will get a stacktrace in your console. There should probably be some kind of exception handling in you code to handle this event 
```
File "/Users/OOH/code/virtualenvs/ert/lib/python3.8/site-packages/dash/dash.py", line 994, in add_context
    output_value = func(*args, **kwargs)  # %% callback invoked %%
  File "/Users/OOH/code/virtualenvs/ert/lib/python3.8/site-packages/ertviz/controllers/observation_response_controller.py", line 160, in _update_graph
    return plot.repr
  File "/Users/OOH/code/virtualenvs/ert/lib/python3.8/site-packages/ertviz/models/plot_model.py", line 141, in repr
    data.append(list(self._data_df_dict[response_name].values.flatten()))
AttributeError: 'NoneType' object has no attribute 'values'

```

File: ertviz/controllers/observation_response_controller.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        name="Observations",
+        **style,
+    )
+    return [observation_data]
+
+
+def _create_misfits_plot(response, selected_realizations, color):
+
+    x_axis = response.axis
+    # realizations = _get_univariate_misfits_plots(
```
Comment (oysteoh) [inline ertviz/controllers/observation_response_controller.py:70]: should be removed if not used

File: ertviz/controllers/observation_response_controller.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    realizations = _get_univariate_misfits_boxplots(
+        response.univariate_misfits_df(selected_realizations),
+        color=color,
+    )
+    ensemble_plot = ResponsePlotModel(
+        realizations,
+        [],
+        dict(
+            xaxis={"title": "Index"},
+            yaxis={"title": "Unit TODO"},
```
Comment (oysteoh) [inline ertviz/controllers/observation_response_controller.py:80]: Probably a name change of the yaxis pending?

File: ertviz/controllers/observation_response_controller.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    return ensemble_plot
+
+
+def observation_response_controller(parent, app):
+    @app.callback(
+        Output(parent.uuid("response-selector"), "options"),
+        [Input(parent.uuid("ensemble-selection-store"), "data")],
+    )
+    def _set_response_options(selected_ensembles):
+        # Should either return a union of all possible responses or the other thing which I cant think of...
```
Comment (oysteoh) [inline ertviz/controllers/observation_response_controller.py:94]: 🤔

File: ertviz/controllers/observation_response_controller.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        if misfits_type == "Summary":
+            data_dict = {}
+            colors = {}
+            for ensemble_id, color in selected_ensembles.items():
+                ensemble = parent.ensembles.get(ensemble_id, None)
+                data_dict[parent.ensembles[ensemble_id]._name] = ensemble.responses[
+                    response
+                ].summary_misfits_df(
+                    selection=None
+                )  # What about selections?
```
Comment (oysteoh) [inline ertviz/controllers/observation_response_controller.py:151]: comment you have left addressing a question to yourself? Is it solved ? should the comment go?

File: ertviz/models/plot_model.py [hunk 6]
```diff
@@ -3,6 +3,41 @@
...
+        self._x_axis = kwargs["x_axis"]
+        self._y_axis = kwargs["y_axis"]
+        self._text = kwargs["text"]
+        self._name = kwargs["name"]
+        self._color = kwargs["color"]
+
+    @property
+    def repr(self):
+        repr_dict = dict(
+            # x=self._x_axis,
```
Comment (oysteoh) [inline ertviz/models/plot_model.py:18]: remove

File: ertviz/models/plot_model.py [hunk 7]
```diff
@@ -3,6 +3,41 @@
...
+            y=self._y_axis,
+            # text=self.display_name,
```
Comment (oysteoh) [inline ertviz/models/plot_model.py:20]: remove

File: ertviz/models/realization.py [hunk 8]
```diff
@@ -1,4 +1,47 @@
...
+        self._data = realization_schema["data"]
+        self._univariate_misfits_df = self._extract_univariate_misfits(
+            realization_schema.get("univariate_misfits")
+        )
+        self._summarized_missfits_value = self._extract_summary_misfits(
+            realization_schema.get("summarized_misfits")
+        )
+
+    def _extract_univariate_misfits(self, schema):
+        if bool(schema):  # this account for not None and empty dict
```
Comment (oysteoh) [inline ertviz/models/realization.py:17]: Often i find it more readable doing the return early and near the check of the variable 
```
if not bool(schema):
    return None

misfits_ = list(schema.values())
..
..
.
return df
```

**Review by oysteoh — CHANGES_REQUESTED** [review]
Had some comments which you can look at, then remember to black your files and squash your commits and you should be good to go 👍  Nice work!

File: ertviz/controllers/observation_response_controller.py (hunk 5 above)
Comment (xjules) [reply ertviz/controllers/observation_response_controller.py:151]: That's comment for us for how to solve selection.

File: ertviz/controllers/observation_response_controller.py (hunk 2 above)
Comment (xjules) [reply ertviz/controllers/observation_response_controller.py:70]: Yes, that was an alternative until we are sure that boxplots is the way. Now we are sure :)

File: ertviz/controllers/observation_response_controller.py (hunk 3 above)
Comment (xjules) [reply ertviz/controllers/observation_response_controller.py:45]: This requires some additional metadata. Removing it for now.

**Comment by xjules** [comment]
> when choosing a response without observations - and then clicking summary you will get a stacktrace in your console. There should probably be some kind of exception handling in you code to handle this event
> 
> ```
> File "/Users/OOH/code/virtualenvs/ert/lib/python3.8/site-packages/dash/dash.py", line 994, in add_context
>     output_value = func(*args, **kwargs)  # %% callback invoked %%
>   File "/Users/OOH/code/virtualenvs/ert/lib/python3.8/site-packages/ertviz/controllers/observation_response_controller.py", line 160, in _update_graph
>     return plot.repr
>   File "/Users/OOH/code/virtualenvs/ert/lib/python3.8/site-packages/ertviz/models/plot_model.py", line 141, in repr
>     data.append(list(self._data_df_dict[response_name].values.flatten()))
> AttributeError: 'NoneType' object has no attribute 'values'
> ```

Should be fixed now

File: ertviz/controllers/observation_response_controller.py (hunk 4 above)
Comment (xjules) [reply ertviz/controllers/observation_response_controller.py:59]: Removed the comment for now as this represents only a list of selected ensembles

File: ertviz/models/plot_model.py (hunk 7 above)
Comment (xjules) [reply ertviz/models/plot_model.py]: Removing for now, left another comment for `x=self._x_axis` which we should discuss on our next presentation.

File: ertviz/models/realization.py (hunk 8 above)
Comment (xjules) [reply ertviz/models/realization.py:17]: will do!

File: ertviz/assets/ert-style.json (hunk 1 above)
Comment (xjules) [reply ertviz/assets/ert-style.json:101]: That's auto-formated with black :shrug:

**Comment by xjules** [comment]
Boxplots of 3 ensembles missfits:
image

File: ertviz/controllers/observation_response_controller.py (hunk 3 above)
Comment (oysteoh) [reply ertviz/controllers/observation_response_controller.py:45]: still not removed?

**Comment by oysteoh** [comment]
Everything seems to work as expected for me now - and no error msgs 👍  with regards to how the plotting is done i'm not sure what would be best - i guess this is something you will figure out / discuss with someone with an opinion! You might also poke Eivind and ask if he has any opinions...

**Review by oysteoh — APPROVED** [review]

**Comment by oysteoh** [comment]
From my point of view it seems ok - but lacks testing. Please squash before merging ⚠️

**Review by oysteoh — APPROVED** [review]

File: ertviz/controllers/observation_response_controller.py (hunk 3 above)
Comment (xjules) [reply ertviz/controllers/observation_response_controller.py]: Should be gone now

File: ertviz/models/plot_model.py (hunk 6 above)
Comment (xjules) [reply ertviz/models/plot_model.py]: I haven't removed it but made a comment that sort of clarifies what's needed to be done

**Pull request merged**